server/client.go

```diff
@@ -3998,8 +3998,9 @@ func (c *client) processMsgResults(acc *Account, r *SublistResult, msg, deliver,
 			// We handle similarly to routes and use the same data structures.
 			// Leaf node delivery audience is different however.
 			// Also leaf nodes are always no echo, so we make sure we are not
-			// going to send back to ourselves here.
-			if c != sub.client && (c.kind != ROUTER || sub.client.isHubLeafNode()) {
+			// going to send back to ourselves here. For messages from routes we want
+			// to suppress in general unless we know from the hub or its a service reply.
+			if c != sub.client && (c.kind != ROUTER || sub.client.isHubLeafNode() || isServiceReply(c.pa.subject)) {
```

Comment:
To minimize the changes and matches the comment, should we restrict the check for isServiceReply to route clients? Something like `... || (c.kind == ROUTER && isServiceReply(c.pa.subject))) {`

Reply:
If it is evaluated (last condition) we already know that the kind is ROUTER no? We even know its a route and not a hub..

Reply:
Right...